### DISCUSSION TITLE
Port proposal: CCBP/Calibrated-Partial-Resets utility-gated pull-toward-init on sigma0_ndecay099 (screen only, no runs yet)

### DISCUSSION THREAD

**Discussion by rama0x1** [SlopDotCash/asi#396] [Ideas]
## Lane, metric, baseline

- Lane: `alberta_framework.benchmarks.ipmnist_screening` (online input-permuted MNIST, `average_online_accuracy`).
- Named incumbent for this comparison: `sigma0_ndecay099` (the frontier-wave champion from `FINAL_REPORT.md`), NOT the current overall campaign champion. The current overall campaign champion is `rls_head_resid_l1_preset005` (0.87114 +/- 0.00010, n=20, per `docs/research/ipmnist-theory.md` / `summary_rls_head_confirm.json`), which uses a streaming-RLS readout architecture this arm's mechanism (a gradient-based per-weight pull toward init) does not compose with directly. This PR is a **mechanism-dissection port**, not a SOTA-climb attempt — same category as the existing `summary_comparison_vs_champion.json` rows (`l2init_ema_norm`, `wclip_ema_norm`, etc.) that are pinned against `sigma0_ndecay099` for the same reason.
- Baseline read from `outputs/ipmnist_screening/summary_comparison_vs_champion.json` and `FINAL_REPORT.md`: `sigma0_ndecay099` 60-task screen mean 0.8616 (seeds 0-2), 200-task confirm mean 0.86245.

## What is changing

One variable: an additive, UPGD-utility-**gated** pull of every hidden parameter toward its value at initialization, layered on top of the `sigma0_ndecay099` champion's existing per-step update (normalize -> UPGD-gated SGD -> decoupled weight decay). Ported mechanism: continuous, per-unit-utility-scaled soft resets (CCBP, OpenReview:UJqXhFFzKu; Calibrated Partial Resets, arXiv:2607.24996). Deviation: additive alongside the champion's existing uniform decay rather than a replacement of it, and reuses the champion's own UPGD utility gate as the per-unit reset weight (1 - gate) since this arm exposes no separate hidden-unit-firing statistic. New registry arm: `sigma0_ndecay099_gated_l2init`, `l2init_pull_scale=0.01` (matches this repo's established L2-Init strength elsewhere, e.g. `upgd_l2init`, `l2init_ema_norm`). `l2init_pull_scale=0` reduces bit-exactly to `sigma0_ndecay099` (unit-test pinned, failing-test-first).

## Seeds, n, threshold

- Tuning: none (pull_scale fixed a priori at the repo's standard L2-Init strength, not tuned on this arm).
- Evaluation/screen: seeds 0, 1, 2 (shared/paired with the existing `sigma0_ndecay099` control shards under `outputs/ipmnist_screening/shards/`), 60 tasks x 5,000 steps, exact per-step noise mode (arm is noise-free, sigma=0).
- Escalation gate (matching this lane's existing comparison-wave convention): paired mean improvement over `sigma0_ndecay099` > 0.002 with every seed positive to flag as a confirmation candidate; otherwise report tie/loss as-is. Given prior L2-Init-family rows in this exact comparison (`l2init_ema_norm` +0.0025, `upgd_l2init` well below the raw-input champion), a small or negative delta is the expected outcome and will be reported plainly either way.

## What I will report if it loses

The measured mean_diff, per-seed diffs, and stderr from `merge --control-name sigma0_ndecay099`, with the arm and its unit tests left in the registry as a documented negative/marginal result (not deleted, since it is a cheap, correctly-reducing registry entry other contributors can build on) unless a maintainer prefers it removed.

Posting before running the screen shards so nobody duplicates this specific comparison.

**Comment by rama0x1** [comment]
Update: the arm is implemented and unit-tested (bit-exact reduction pin at `l2init_pull_scale=0`, hand-computed-trajectory check, finite-smoke check, no-extra-randomness check — all passing), PR up at https://github.com/elizaOS/asi/pull/405.

I could **not** produce the 60-task/seeds-0-2 screening numbers promised above in my execution environment: `sklearn.datasets.fetch_openml("mnist_784", ...)` fails there with a consistent `HTTP Error 504: Gateway Timeout` from `openml.org`'s data-list API (reproduced directly with `curl` against both `www.openml.org` and `api.openml.org`, redirect followed, 504 both times, ~15-16s each), not a local/sandbox network block. So the registered arm's real paired-seed comparison against `sigma0_ndecay099` is still unrun.

The PR therefore only claims "arm registered + bit-exact reduction pinned," not a win, tie, or loss — no benchmark number is asserted. Whoever next has a working path to the canonical MNIST cache can run:

```
.venv/bin/python -m alberta_framework.benchmarks.ipmnist_screening run \
  --config-name sigma0_ndecay099_gated_l2init --seed {0,1,2} --n-tasks 60 --task-length 5000 \
  --out outputs/ipmnist_screening/shards/sigma0_ndecay099_gated_l2init_seed{0,1,2}.json

.venv/bin/python -m alberta_framework.benchmarks.ipmnist_screening merge \
  --shards outputs/ipmnist_screening/shards/sigma0_ndecay099_gated_l2init_seed*.json \
           outputs/ipmnist_screening/shards/sigma0_ndecay099_seed*.json \
  --control-name sigma0_ndecay099 \
  --output outputs/ipmnist_screening/summary_gated_l2init.json
```

and report `paired_vs_control.mean_diff` / `per_seed_diff` / `all_seeds_improve` here against the +0.002-all-positive gate from the pre-registration above.

**Comment by startrack3r** [comment]
I am taking the unrun screen on the current `SlopDotCash/asi` main head
`c9aba7b54dedd647f8bd5f5c7bf6780b1413b676`.

The canonical OpenML MNIST materialization now loads and validates locally as
`(60000, 784) float32` inputs and `(60000,) int32` labels. The focused
`TestGatedL2Init` panel passes: 5 passed.

One prospective protocol tightening before any result exists: current
`docs/research/ipmnist-theory.md` requires the control to be remeasured in the
same runner. I will therefore run both `sigma0_ndecay099` and
`sigma0_ndecay099_gated_l2init` on paired seeds 0, 1, and 2, rather than import
the historical control shards. The published +0.002 paired-mean and
all-seeds-positive gate is unchanged. All results remain development-only and
permanently nonpromoting.

Exact run shape for each arm/seed:

```text
.venv/Scripts/python.exe -m alberta_framework.benchmarks.ipmnist_screening run \
  --config-name <arm> --seed <seed> --n-tasks 60 --task-length 5000 \
  --data-home outputs/upgd_ipmnist/openml_cache --noise-mode step \
  --progress-every 10 \
  --out outputs/ipmnist_screening/gated_l2init_screen_c9aba7b_20260826/shards/<arm>_seed<seed>.json
```

The merge will use only those six same-runner shards, with control
`sigma0_ndecay099`, and will write a new summary in that namespace. I will post
the result here even if it loses. No benchmark shard has been started yet.

**Comment by startrack3r** [comment]
Follow-up: the pre-registered run exposed a Windows publication bug before any candidate arm ran. All three control computations reached final linking, but cleanup of the read-only temporary hard link returned WinError 5. I stopped the comparison and am not treating any partial output as a result. The fix and regression coverage are in https://github.com/SlopDotCash/asi/pull/2508. If it lands, both arms must restart from scratch at the new commit in a new output namespace. No benchmark metric is claimed here.
